# Analyzing Used Car Listings on eBay Kleinanzeigen

The dataset was originally scraped and uploaded to Kaggle. The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
#reading file 
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

### Apart from that, It can be observed that:

- Most of the columns are of object type.
- There are some missing values in the few of the columns such as vehicleType, model, notRepairedDamage
- Odomter values can be converted to numeric type to analyse that column
- Price values can also be converted to numeric type to analyse the price column
- Some of the data needs to converted to english language for better understandablity
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

## Clean Columns

Lest start working on the dataset by changing the column names, refactoring to python's preferred snake case.

In [4]:
autos_colums = autos.columns
print(autos_colums)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
autos_colums = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [6]:
autos.columns = autos_colums

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  year_of_registration  \
0  privat    Angebot  $5,000  control          bus                  2004   
1  privat    Angebot  $8,500  control    limousine                  1997   
2  privat    Angebot  $8,990     test    limousine                  2009   
3  privat    Angebot  $4,350  control   kleinwagen                  2007   
4  privat    Angebot  $1,350     test        kombi                  2003   

     gearbox  power_ps   model   odometer  month_of_registration fuel_type  \
0    manuell       158  andere  150,000km                      3       lpg   
1  automatik       286     7er  150,000km                      6    benzin   
2    manuell       102    golf   70,000km                      7    benzin   
3  automatik        71  fortwo   70,000km                      6    benzin   
4    manuell         0   focus  150,000km                      7    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Data Exploration and Cleaning

Now we have proper named colums in the data set.

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for: 
- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [8]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  year_of_registration  gearbox      power_ps  model  \
count         44905          50000.000000    47320  50000.000000  47242   
unique            8                   NaN        2           NaN    245   
top       limousine                   NaN  manuell           NaN   golf   
freq          12859                   NaN    36993           NaN   4024   
mean            NaN           2005.073280      NaN    116.355920    NaN   
std             NaN            105.712813      NaN    209.216627    NaN   
min             NaN           1000.000000      NaN      0.000000    NaN   
25%             NaN           1999.000000      NaN     70.000000    NaN   
50%             NaN           2003.000000      NaN    105.000000    NaN   
75%             NaN           2008.000000      NaN    150.000000    NaN   
max             NaN           9999.000000      NaN  17700.000000    NaN   

         odometer  month_of_registration fuel_type       brand  \
count       50000           50000.000000     45518       50000   
unique         13                    NaN         7          40   
top     150,000km                    NaN    benzin  volkswagen   
freq        32424                    NaN     30107       10687   
mean          NaN               5.723360       NaN         NaN   
std           NaN               3.711984       NaN         NaN   
min           NaN               0.000000       NaN         NaN   
25%           NaN               3.000000       NaN         NaN   
50%           NaN               6.000000       NaN         NaN   
75%           NaN               9.000000       NaN         NaN   
max           NaN              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Initial Observations:

- There are a number of text columns where all (or nearly all) of the values are the same:
 - seller
 - offer_type
- The num_photos column looks odd, we'll need to investigate this further.
- Price and odometer columns needs to be changed to numberic types ,so that we can further analyse those columns.
- Change the date_crawled and date_created coluns to date time format.

Creating one more column in the autos dataset which will store intiger values of price in dollars.

In [9]:
autos.rename({"nr_of_pictures":"num_photos"}, axis=1, inplace= True)

In [10]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

Looks like num_phots has 0 value in every column, which is not required. We can drop it.

Lets check seller and offer_type

In [11]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

seller column has two different values ie privat and gewerblich , which are not required in our case. 
We can go ahead and drop that column.

Lets check offer_type:

In [12]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

seller column has two different values ie Angebot and Genuch , which are not required in our case.
We can go ahead and drop that column.

In [13]:
autos = autos.drop(["offer_type", "seller", "num_photos"], axis = 1)

In [14]:
autos["price_in_dollar"] = autos["price"].str.replace("$", "").str.replace(",","").astype(int)

In [15]:
autos["price_in_dollar"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_in_dollar, dtype: int64

Converting odometer column values to numeric types.

Creating one more column in the autos dataset which will store intiger values of odometer in Km.

In [16]:
autos["odometer_km"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

In [17]:
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

### Converting `date_crawled` and `date_created` to date time object.

In [19]:
# dates = pd.to_datetime(dates) 
autos["date_created"] = pd.to_datetime(autos.date_created)
autos["date_crawled"] = pd.to_datetime(autos.date_crawled)

## Exploring Odometer and Price

In [24]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

We can see that values in this field are rounded.
Apart from that there are more vehicles with high mileage than low mileage ones.

In [27]:
print(autos["price_in_dollar"].unique().shape)
print(autos["price_in_dollar"].describe())
autos["price_in_dollar"].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_in_dollar, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price_in_dollar, dtype: int64

We can observe that there values are also rounded, my fixed input form.
There are **2357** unique values in this column, out of which are 1421 rows with **$0** price value. 

One more observation is maximum price is 100 million dollars, which looks like a lot, lets explore it furthur.

In [31]:
autos["price_in_dollar"].value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_in_dollar, dtype: int64

In [32]:
autos["price_in_dollar"].value_counts().sort_index(ascending = True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_in_dollar, dtype: int64

There are a number of listings with prices below \$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

We will remove anything beyond price of **$350,000** since it seems unrealistic.

In [34]:
autos = autos[autos["price_in_dollar"].between(1,351000)]

date_crawled                                               name  \
0     2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1     2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2     2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3     2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4     2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5     2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6     2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7     2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8     2016-03-22 16:51:34                                         Seat_Arosa   
9     2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10    2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11    2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12    2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13    2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14    2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15    2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16    2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17    2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18    2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19    2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20    2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21    2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22    2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23    2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24    2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25    2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26    2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28    2016-03-19 21:56:19                                      MINI_Cooper_D   
29    2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30    2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
...                   ...                                                ...   
49968 2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969 2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970 2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971 2016-03-29 14:54:12                                         W.Lupo_1.0   
49972 2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973 2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975 2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976 2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977 2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978 2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979 2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980 2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981 2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982 2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983 2016-03-06 12:43:04                                      Ford_focus_99   
49985 2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986 2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987 2016-03-22 20:47:27  Audi_A3_Limousine_2.0_TDI_DPF_Ambition__NAVI__...   
49988 2016-03

In [36]:
autos.rename({"month_of_registration":"registration_month"},axis =1 , inplace= True)
autos.rename({"year_of_registration":"registration_year"},axis =1 , inplace= True)

## Observing  the date columns
There are a number of columns with date information:
- date_crawled
- registration_month
- registration_year
- ad_created
- last_seen

We have already converted date_crawled and ad_created in datetime format, lets convert last_seen in datetime format.
- registration_month and registration_year are in numeric types so we can exlpore them straight forward.

In [41]:
autos["registration_month"].value_counts().sort_index(ascending = False)

12    3447
11    3360
10    3651
9     3389
8     3191
7     3949
6     4368
5     4107
4     4102
3     5071
2     3008
1     3282
0     5075
Name: registration_month, dtype: int64

In [45]:
autos["registration_year"].value_counts().sort_index(ascending = False).head(20)

9999       4
9996       1
9000       2
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       3
2018     492
2017    1453
2016    1316
2015     399
2014     666
2013     806
2012    1323
2011    1634
Name: registration_year, dtype: int64

In [50]:

print(autos["registration_year"].describe())
autos["registration_year"].unique()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


array([2004, 1997, 2009, 2007, 2003, 2006, 1995, 1998, 2000, 2017, 2010,
       1999, 1982, 1990, 2015, 2014, 1996, 1992, 2005, 2002, 2012, 2011,
       2008, 1985, 2016, 1994, 1986, 2001, 2018, 2013, 1972, 1993, 1988,
       1989, 1967, 1973, 1956, 1976, 4500, 1987, 1991, 1983, 1960, 1969,
       1950, 1978, 1980, 1984, 1963, 1977, 1961, 1968, 1934, 1965, 1971,
       1966, 1979, 1981, 1970, 1974, 1910, 1975, 5000, 4100, 2019, 1959,
       9996, 9999, 6200, 1964, 1958, 1800, 1948, 1931, 1943, 9000, 1941,
       1962, 1927, 1937, 1929, 1000, 1957, 1952, 1111, 1955, 1939, 8888,
       1954, 1938, 2800, 5911, 1500, 1953, 1951, 4800, 1001])

We can notice from above data that there error in the registration_year column.
- some cars are listed in year 1000
- Max year is 9999

This clearly says that most of the registration_years are irrrelavant.
We will only keep the rows with registration years from **1950 to 2019.**

In [51]:
# autos = autos[autos["price_in_dollar"].between(1,351000)]
autos = autos[autos["registration_year"].between(1950,2019)]

In [54]:
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067144
2005    0.060358
1999    0.060058
2004    0.054793
2003    0.054592
2006    0.054212
2001    0.054112
2002    0.050709
1998    0.049107
2007    0.046124
Name: registration_year, dtype: float64

We can see that most of the cars were registered in 1990's.

## Exploring Price by Brand

In [56]:
autos["brand"].value_counts(normalize= True)

volkswagen        0.213825
opel              0.109205
bmw               0.108644
mercedes_benz     0.094691
audi              0.085742
ford              0.069587
renault           0.048106
peugeot           0.029148
fiat              0.026185
seat              0.018818
skoda             0.015715
mazda             0.015155
nissan            0.015094
smart             0.014033
citroen           0.014013
toyota            0.012352
sonstige_autos    0.010610
hyundai           0.009769
volvo             0.009149
mini              0.008488
mitsubishi        0.008088
honda             0.007988
kia               0.007127
alfa_romeo        0.006586
porsche           0.005886
suzuki            0.005866
chevrolet         0.005665
chrysler          0.003623
dacia             0.002582
daihatsu          0.002562
jeep              0.002182
subaru            0.002162
land_rover        0.001982
saab              0.001582
daewoo            0.001582
jaguar            0.001541
trabant           0.001521
r

**Volkswagen is the most sold where are some brands have extremly low ratings. For ex: lada,lancia**

We will limit our analysis to brands representing more than 5% of listings.

In [60]:
brand_counts = autos["brand"].value_counts(normalize= True)
common_brands = brand_counts[brand_counts > 0.05].index
print(common_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [63]:
brand_men_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    brand_mean = brand_only["price_in_dollar"].mean()
    brand_men_prices[brand] = int(brand_mean)
print(brand_men_prices)

{'volkswagen': 6382, 'opel': 5106, 'bmw': 8253, 'mercedes_benz': 29517, 'audi': 8965, 'ford': 7097}


**All the 5 top brands have huge price difference**
- cheapes one being opel and expensive one is mercedes_benz
- Audi and Ford are among top 3 expensive ones.

## Exploring Mileage

In [64]:
bmp_series = pd.Series(brand_men_prices)
pd.DataFrame(bmp_series,columns=["mean_price"])

mean_price
volkswagen           6382
opel                 5106
bmw                  8253
mercedes_benz       29517
audi                 8965
ford                 7097

In [67]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_men_prices).sort_values(ascending=False)

In [69]:
brand_info = pd.DataFrame(mean_mileage, columns=["mean_mileage"])

In [70]:
brand_info["mean_prices"] = mean_prices

In [71]:
brand_info

mean_mileage  mean_prices
bmw                  132567         8253
mercedes_benz        130978        29517
audi                 129643         8965
opel                 129389         5106
volkswagen           129015         6382
ford                 124233         7097

From above data frame we can say that expensive brands have higher mileage than the cheap brands.